advanced_source/ddp_pipeline_tutorial.py

```diff
@@ -0,0 +1,455 @@
+            # Need to move targets to the device where the output of the
+            # pipeline resides.
+            loss = criterion(output.view(-1, ntokens), targets.cuda(2 * rank + 1))
+            loss.backward()
+            torch.nn.utils.clip_grad_norm_(model.parameters(), 0.5)
+            optimizer.step()
+
+            total_loss += loss.item()
+            log_interval = 10
+            if batch % log_interval == 0 and batch > 0:
+                cur_loss = total_loss / log_interval
+                elapsed = time.time() - start_time
+                print_with_rank('| epoch {:3d} | {:5d}/{:5d} batches | '
+                      'lr {:02.2f} | ms/batch {:5.2f} | '
+                      'loss {:5.2f} | ppl {:8.2f}'.format(
+                        epoch, batch, nbatches // bptt, scheduler.get_lr()[0],
+                        elapsed * 1000 / log_interval,
+                        cur_loss, math.exp(cur_loss)))
+                total_loss = 0
+                start_time = time.time()
+
+    def evaluate(eval_model, data_source):
+        eval_model.eval() # Turn on the evaluation mode
+        total_loss = 0.
+        ntokens = len(TEXT.vocab.stoi)
+        # Evaluate only for 50 batches to keep script execution time low.
+        nbatches = min(50 * bptt, data_source.size(0) - 1)
+        with torch.no_grad():
+            for i in range(0, nbatches, bptt):
+                data, targets = get_batch(data_source, i)
+                output = eval_model(data).local_value()
+                output_flat = output.view(-1, ntokens)
+                # Need to move targets to the device where the output of the
+                # pipeline resides.
+                total_loss += len(data) * criterion(output_flat, targets.cuda(2 * rank + 1)).item()
+        return total_loss / (len(data_source) - 1)
+
+######################################################################
+# Loop over epochs. Save the model if the validation loss is the best
+# we've seen so far. Adjust the learning rate after each epoch.
+
+# In 'run_worker'
+    best_val_loss = float("inf")
+    epochs = 3 # The number of epochs
+    best_model = None
+
+    for epoch in range(1, epochs + 1):
+        epoch_start_time = time.time()
+        train()
+        val_loss = evaluate(model, val_data)
+        print_with_rank('-' * 89)
+        print_with_rank('| end of epoch {:3d} | time: {:5.2f}s | valid loss {:5.2f} | '
+              'valid ppl {:8.2f}'.format(epoch, (time.time() - epoch_start_time),
+                                         val_loss, math.exp(val_loss)))
+        print_with_rank('-' * 89)
+
+        if val_loss < best_val_loss:
+            best_val_loss = val_loss
+            best_model = model
+
+        scheduler.step()
+
+
+######################################################################
+# Evaluate the model with the test dataset
+# -------------------------------------
+#
+# Apply the best model to check the result with the test dataset.
```

Comment:
Could we find best model across replicas? Say do an all-reduce and print only if the model has best loss.

Reply:
As I understand it, they should be equivalent right? With DDP they should start off with the same params and gradients are synced every iteration.

Reply:
I see, although each of them my come to different val_loss and best_val_loss, but I see at the end it should not matter which one you choose.